category/vm/compiler/ir/x86/emitter.cpp

```diff
@@ -1009,22 +1009,17 @@ namespace monad::vm::compiler::native
         return block_prologue(b);
     }
 
-    void Emitter::gas_decrement_no_check(int32_t gas)
+    void Emitter::gas_decrement_static_work(int32_t gas)
     {
-        as_.sub(
-            x86::qword_ptr(reg_context, runtime::context_offset_gas_remaining),
-            gas);
-    }
-
-    void Emitter::gas_decrement_no_check(asmjit::x86::Gpq reg)
-    {
-        as_.sub(
-            x86::qword_ptr(reg_context, runtime::context_offset_gas_remaining),
-            reg);
+        gas_decrement_no_check(gas);
+        if (!accumulate_static_work(gas)) {
+            as_.jl(error_label_);
+        }
     }
 
-    void Emitter::gas_decrement_check_non_negative(int32_t gas)
+    void Emitter::gas_decrement_unbounded_work(int32_t gas)
     {
+        accumulated_static_work_ = 0;
         gas_decrement_no_check(gas);
         as_.jl(error_label_);
```

Comment:
If `gas == 0`, `gas_decrement_no_check` wont set the flags and this conditional jump will use the flags from the previous instruction.